src/main/java/com/yelp/nrtsearch/server/luceneserver/field/IndexableFieldDef.java

```diff
@@ -0,0 +1,268 @@
+import java.io.IOException;
+import java.util.List;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.similarities.BM25Similarity;
+import org.apache.lucene.search.similarities.ClassicSimilarity;
+import org.apache.lucene.search.similarities.Similarity;
+
+/** Base class for all field definition types that can be written into the index. */
+public abstract class IndexableFieldDef extends FieldDef {
+  public static final Similarity DEFAULT_SIMILARITY = new BM25Similarity();
+
+  public enum FacetValueType {
+    NO_FACETS,
+    FLAT,
+    HIERARCHY,
+    NUMERIC_RANGE,
+    SORTED_SET_DOC_VALUES
+  }
+
+  private final boolean isStored;
+  private final boolean isMultiValue;
+  private final boolean isSearchable;
+  private final String postingsFormat;
+  private final String docValuesFormat;
+  private final Similarity similarity;
+
+  protected final DocValuesType docValuesType;
+  protected final IndexableFieldDef.FacetValueType facetValueType;
+  protected final FieldType fieldType;
+
+  /**
+   * Field constructor. Performs generalized building of field definition by calling a number of
+   * protected methods, mainly: {@link #validateRequest(Field)}, {@link #parseDocValuesType(Field)},
+   * {@link #parseFacetValueType(Field)}, and {@link #setSearchProperties(FieldType, Field)}.
+   * Concrete field types should override these methods as needed.
+   *
+   * @param name name of field
+   * @param requestField field definition from grpc request
+   */
+  protected IndexableFieldDef(String name, Field requestField) {
+    super(name);
+
+    validateRequest(requestField);
+    isStored = requestField.getStore();
+    isMultiValue = requestField.getMultiValued();
+    isSearchable = requestField.getSearch();
+    docValuesType = parseDocValuesType(requestField);
+
+    fieldType = new FieldType();
+    fieldType.setStored(requestField.getStore());
+    setSearchProperties(fieldType, requestField);
+    fieldType.freeze();
+
+    facetValueType = parseFacetValueType(requestField);
+
+    postingsFormat =
+        requestField.getPostingsFormat().isEmpty()
+            ? ServerCodec.DEFAULT_POSTINGS_FORMAT
+            : requestField.getPostingsFormat();
+    docValuesFormat =
+        requestField.getDocValuesFormat().isEmpty()
+            ? ServerCodec.DEFAULT_DOC_VALUES_FORMAT
+            : requestField.getDocValuesFormat();
+
+    String similarityStr = requestField.getSimilarity();
+    if (!similarityStr.isEmpty()) {
+      if (similarityStr.equalsIgnoreCase("DefaultSimilarity")) {
+        this.similarity = new ClassicSimilarity();
+      } else if (similarityStr.equalsIgnoreCase("BM25Similarity")) {
+        // TODO pass custom k1, b values so that we can call Constructor BM25Similarity(float k1,
+        // float b)
+        this.similarity = new BM25Similarity();
```

Comment:
Do we want `Postings` and `Similarity` in `IndexableFieldDef`? Since these might only be needed for text fields, could we move them to `TextBaseFieldDef`?

Reply:
I wasn't entirely sure the best level to put these properties. But if that is the case, I can move them.

Reply:
We could also do it in a subsequent review. Closing for now.

Reply:
Yeah. I was going to move it quickly, but boolean needs posting because it is indexed. I think we should fix boolean first and sort this out later.